log/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+            }
+            _ => (feature as *const lv2_sys::LV2_Log_Log)
+                .as_ref()
+                .map(|internal| Self { internal }),
+        }
+    }
+}
+
+impl<'a> Log<'a> {
+    /// Send a log message to the host.
+    ///
+    /// The `entry_type` parameter is an URID representing the kind of log message. There are four
+    /// kind of message:
+    /// * **note:** an informative message.
+    /// * **warning:** a warning message.
+    /// * **error:** an error message.
+    /// * **trace:** a debugging trace. These entries should not be displayed during normal
+    /// operation, but the host may implement an option to display them for debugging purposes.
+    /// This entry type is special in that it may be written to in a real-time thread. It is
+    /// assumed that if debug tracing is enabled, real-time considerations are not a concern.
+    pub fn print(&self, entry_type: URID<impl EntryType>, message: &str) -> Result<(), LogError> {
+        let printf = if let Some(printf) = self.internal.printf {
+            printf
+        } else {
+            return Err(LogError::NoCallback);
+        };
+        let message = String::from(message) + "\0";
+
+        let res = unsafe {
+            (printf)(
+                self.internal.handle,
+                entry_type.get(),
+                "%s\0" as *const _ as *const c_char,
+                message.as_str() as *const _ as *const c_char,
+            )
+        };
+        if res > 0 {
+            Ok(())
+        } else {
+            Err(LogError::PrintError)
+        }
+    }
+    /// Send a message to the host with a new line at the end.
+    ///
+    /// It same as [print](struct.Log.html#method.print) but add a newline a the end of message.
+    /// See [print](struct.Log.html#method.print) documentation for details.
+    pub fn println(&self, entry_type: URID<impl EntryType>, message: &str) -> Result<(), LogError> {
+        let printf = if let Some(printf) = self.internal.printf {
+            printf
+        } else {
+            return Err(LogError::NoCallback);
+        };
+        let message = String::from(message) + "\n\0";
+
+        let res = unsafe {
+            (printf)(
+                self.internal.handle,
+                entry_type.get(),
+                "%s\0" as *const _ as *const c_char,
+                message.as_str() as *const _ as *const c_char,
+            )
+        };
+        if res > 0 {
+            Ok(())
+        } else {
+            Err(LogError::PrintError)
+        }
+    }
+}
+
+/// A URID cache containing all log properties.
+#[derive(URIDCollection, Debug)]
+pub struct LogURIDCollection {
+    pub entry_class: URID<EntryClass>,
+    pub error_class: URID<ErrorClass>,
+    pub note_class: URID<NoteClass>,
+    pub trace_class: URID<TraceClass>,
+    pub warning_class: URID<WarningClass>,
+    //pub log: URID<Log<'a>>,
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
There should probably be some tests in there (ideally in another file in the `tests` directory), just to make sure the API is at least functional. :slightly_smiling_face:

Reply:
Is it enought to check if a basic example can compile ? In this project it's frequently more difficult to setting up everything for the test than writing an actual plugin.